netZooPy/panda/panda.py

```diff
@@ -420,32 +420,68 @@ def update_diagonal(diagonal_matrix, num, alpha, step):
             diagonal_fill = diagonal_std * num * math.exp(2 * alpha * step)
             np.fill_diagonal(diagonal_matrix, diagonal_fill)
 
+        def gt_function(x, y=None):
+            '''T function.'''
+            if y is None:
+                a_matrix = cp.dot(x, x.T)
+                s = cp.square(x).sum(axis=1)
+                a_matrix /= cp.sqrt(s + s.reshape(-1, 1) - cp.abs(a_matrix))
+            else:
+                a_matrix = cp.dot(x, y)
+                a_matrix /= cp.sqrt(cp.square(y).sum(axis=0) + cp.square(x).sum(axis=1).reshape(-1, 1) - cp.abs(a_matrix))
+            return a_matrix
+
+        def gupdate_diagonal(diagonal_matrix, num, alpha, step):
+            '''Update diagonal.'''
+            cp.fill_diagonal(diagonal_matrix, cp.nan)
+            diagonal_std = cp.nanstd(diagonal_matrix, 1)
+            diagonal_fill = diagonal_std * num * math.exp(2 * alpha * step)
+            cp.fill_diagonal(diagonal_matrix, diagonal_fill)
+
         panda_loop_time = time.time()
         num_tfs, num_genes = motif_matrix.shape
         step = 0
         hamming = 1
         alpha = 0.1
+        
         while hamming > 0.001:
             # Update motif_matrix
-            W = 0.5 * (t_function(ppi_matrix, motif_matrix) + t_function(motif_matrix, correlation_matrix))  # W = (R + A) / 2
-            hamming = np.abs(motif_matrix - W).mean()
-            motif_matrix *= (1 - alpha)
-            motif_matrix += (alpha * W)
-
-            if hamming > 0.001:
```

Comment:
@dcolinmorgan you delete this line (435) in the original code but it was not replaced in your PR